cpp/src/arrow/array/builder_nested.h

```diff
@@ -80,100 +89,91 @@ class BaseListBuilder : public ArrayBuilder {
     value_builder_->Reset();
   }
 
-  /// \brief Vector append
-  ///
-  /// If passed, valid_bytes is of equal length to values, and any zero byte
-  /// will be considered as a null for that slot
-  Status AppendValues(const offset_type* offsets, int64_t length,
-                      const uint8_t* valid_bytes = NULLPTR) {
-    ARROW_RETURN_NOT_OK(Reserve(length));
-    UnsafeAppendToBitmap(valid_bytes, length);
-    offsets_builder_.UnsafeAppend(offsets, length);
-    return Status::OK();
-  }
-
   /// \brief Start a new variable-length list slot
   ///
   /// This function should be called before beginning to append elements to the
-  /// value builder
-  Status Append(bool is_valid = true) {
+  /// value builder.
+  ///
+  /// \pre if is_valid is false, list_length MUST be 0
+  /// \param is_valid Whether the new list slot is valid
+  /// \param list_length The number of elements in the list
+  Status Append(bool is_valid, int64_t list_length) {
```

Comment:
Is it really necessary to have the caller pass the list length, or could it simply inferred in the next `Append` or `Finish` call?

Reply:
I want the list-view builder to allow users to control the sizes they push into the sizes buffer.

Example: I can append logical values `[4, 5, 6]`, `[4, 5]`, and `[4]` by

```
Append(true, 3)
Append(true, 2)
Append(true, 1)
child->Append([4, 5, 6])
```

List of suffixes is also possible with careful alternations of child mutations and size appends.

Reply:
Not that this is the new parent class of both list and list-view builders. The classic list builders remain with their API `Append(bool)` is still supported.

Reply:
The builders are mirroring the same sub-typing chain I created for the Arrays.

![image](https://github.com/apache/arrow/assets/207795/0603f862-550a-484d-8940-082122ff5181)

Reply:
This would be useful, can we add this in comment? Since the interface would be a bit tricky.